test/resource.cpp

```diff
@@ -13,70 +13,149 @@
+#ifdef CAMP_HAVE_CUDA
+  test_copy<Cuda>();
+#endif
+#ifdef CAMP_HAVE_HIP
+  test_copy<Hip>();
+#endif
+#ifdef CAMP_HAVE_OMP_OFFLOAD
+  test_copy<Omp>();
+#endif
+#ifdef CAMP_HAVE_SYCL
+  test_copy<Sycl>();
+#endif
 }
-TEST(CampResource, ConvertFails)
+
+template < typename Res, typename Res2 >
+void test_convert_fails()
 {
-  Resource h1{Host()};
-  h1.get<Host>();
-  ASSERT_THROW(h1.get<Host2>(), std::runtime_error);
-  ASSERT_FALSE(h1.try_get<Host2>());
+  Resource r{Res()};
+  r.get<Res>();
+  ASSERT_THROW(r.get<Res2>(), std::runtime_error);
+  ASSERT_FALSE(r.try_get<Res2>());
 }
-TEST(CampResource, GetPlatform)
+//
+TEST(CampResource, ConvertFails)
 {
-  ASSERT_EQ(static_cast<const Resource>(Host()).get_platform(), Platform::host);
+  test_convert_fails<Host, Host2>();
 #ifdef CAMP_HAVE_CUDA
-  ASSERT_EQ(static_cast<const Resource>(Cuda()).get_platform(), Platform::cuda);
+  test_convert_fails<Cuda, Cuda2>();
 #endif
 #ifdef CAMP_HAVE_HIP
-  ASSERT_EQ(static_cast<const Resource>(Hip()).get_platform(), Platform::hip);
+  test_convert_fails<Hip, Hip2>();
 #endif
 #ifdef CAMP_HAVE_OMP_OFFLOAD
-  ASSERT_EQ(static_cast<const Resource>(Omp()).get_platform(), Platform::omp_target);
+  test_convert_fails<Omp, Omp2>();
+#endif
+#ifdef CAMP_HAVE_SYCL
+  test_convert_fails<Sycl, Sycl2>();
 #endif
 }
-TEST(CampResource, Compare)
+
+template < typename Res >
+void test_convert_works(Platform platform)
 {
-  Resource h1{Host()};
-  Resource h2{Host()};
-  Host h; Resource h3{h};
+  Resource r{Res()};
+  ASSERT_TRUE(r.try_get<Res>());
+  ASSERT_EQ(r.get<Res>().get_platform(), platform);
+}
+//
+TEST(CampResource, ConvertWorks)
+{
+  test_convert_works<Host>(Platform::host);
+#ifdef CAMP_HAVE_CUDA
+  test_convert_works<Cuda>(Platform::cuda);
+#endif
+#ifdef CAMP_HAVE_HIP
+  test_convert_works<Hip>(Platform::hip);
+#endif
+#ifdef CAMP_HAVE_OMP_OFFLOAD
+  test_convert_works<Omp>(Platform::omp_target);
+#endif
+#ifdef CAMP_HAVE_SYCL
+  test_convert_works<Sycl>(Platform::sycl);
+#endif
+}
 
+TEST(CampResource, GetPlatform)
+{
+  ASSERT_EQ(static_cast<const Resource>(Host()).get_platform(), Platform::host);
 #ifdef CAMP_HAVE_CUDA
-  Resource r1{Cuda()};
-  Resource r2{Cuda()};
-  Cuda s; Resource r3{s};
+  ASSERT_EQ(static_cast<const Resource>(Cuda()).get_platform(), Platform::cuda);
 #endif
 #ifdef CAMP_HAVE_HIP
-  Resource r1{Hip()};
-  Resource r2{Hip()};
-  Hip s; Resource r3{s};
+  ASSERT_EQ(static_cast<const Resource>(Hip()).get_platform(), Platform::hip);
 #endif
 #ifdef CAMP_HAVE_OMP_OFFLOAD
-  Resource r1{Omp()};
-  Resource r2{Omp()};
-  Omp s; Resource r3{s};
+  ASSERT_EQ(static_cast<const Resource>(Omp()).get_platform(), Platform::omp_target);
 #endif
+#ifdef CAMP_HAVE_SYCL
+  ASSERT_EQ(static_cast<const Resource>(Sycl()).get_platform(), Platform::sycl);
+#endif
+}
 
-  ASSERT_TRUE(h1 == h1);
-  ASSERT_TRUE(h1 == h2);
-  
-  ASSERT_FALSE(h1 != h2);
+template < typename Res >
+void test_compare(Resource& h1, Resource& h2, Resource& h3)
```

Comment:
This test is hard for me to follow. Is it testing the following cases?
- Self equality (but then why no r3 == r3 or r == r3?)
- Inequality between different objects of the same resource type (but then why no r1 != r3 or r2 != r3 and the reverse ordering?)
- Inequality between different resource types (but then why no r2 != h2 or checks of the reverse ordering?)

Reply:
Its probably also helpful to know that all Host resources compare == to all other Host resources.